packages/venia-concept/src/ServiceWorker/Utilities/htmlHandler.js

```diff
@@ -142,17 +143,75 @@ const cloneRequestWithDiffURL = (request, url) =>
           })
         : request;
 
+/**
+ * Function validates if the script files mentioned in the cached
+ * HTML file are cached. If not, it returns false.
+ *
+ * @returns Promise<Boolean>
+ */
+const areAllHTMLLinksCached = async () => {
+    const cachedResponseObj = await caches.match('/');
+
+    if (cachedResponseObj) {
+        const cachedResponse = await cachedResponseObj.text();
+        const parsedCacheResponse = parse(cachedResponse, {
+            style: true
+        });
+        const scriptFiles = parsedCacheResponse
+            .querySelectorAll('script')
+            .map(generateScriptResource)
+            .filter(identity);
+
+        const fileExistsArray = await Promise.all(
+            scriptFiles.map(file => caches.match(file))
+        );
+
+        return fileExistsArray.every(x => !!x);
+    } else {
+        return Promise.resolve(false);
+    }
+};
+
+/**
+ * Function will bust the HTML file cached in the storage.
+ *
+ * @returns Promise
+ */
+const bustHTMLCache = async () => {
+    const runtimeCache = await caches.open(RUNTIME_CACHE_NAME);
+
+    return runtimeCache.delete('/');
+};
+
 /**
  * cacheHTMLPlugin is a workbox plugin that will apply request
  * and reponse manipulations on HTML routes.
  */
 export const cacheHTMLPlugin = {
-    cacheKeyWillBeUsed: async () => {
+    cacheKeyWillBeUsed: ({ mode }) => {
         /**
-         * For all read and writes that arrive on any HTML route
-         * return `/`.
+         * In read mode,
+         *
+         * 1. Check if all the script links mentioned in the
+         * cached HTML are cached.
+         * 2. If after step 1, we realize all the links are not
+         * cached, delete HTML cache, and return '/'
+         * 3. If after step 1, we realize all the links are cached,
+         * we do not delete the cache, we can use the cached HTML.
          */
-        return '/';
+        if (mode === 'read') {
+            return areAllHTMLLinksCached().then(allLinksCached => {
```

Comment:
I was compelled to use `.then` instead of `async-await` here. I want `return '/'` to only happen after `areAllHTMLLinksCached` and `bustHTMLCache` if `areAllHTMLLinksCached` returned `true`. Not sure how to achieve that with `async-await`. It was returning early because `return '/'` is not dependent on the results of the above 2 `await` functions. Let me know if I can achieve this with `async-await`. This is a version of what I had previously: 

```es6
const allLinksCached = await areAllHTMLLinksCached();

if (!allLinksCached) {
  await bustHTMLCache();
}

return Promise.resolve('/')
```

Reply:
I think I've got everything working, only difference is removing the unneeded `Promise.resolve`, not sure that would have caused your code not to work though. Does this work for you?

```
cacheKeyWillBeUsed: async ({ mode }) => {
    if (mode === 'read') {
        const allLinksCached = await areAllHTMLLinksCached();
        if (!allLinksCached) {
            await bustHTMLCache();
        }
    }

    return '/';
}
```